### PR TITLE
HPackDecoder performance

### PR DESCRIPTION
**Pull request by JamesNK**
Addresses https://github.com/dotnet/runtime/issues/1578

Changes of note:
1. Avoids the entry state machine where possible. Parsing strings and numbers would go back to the entry state machine with every byte. Now loops are much tighter (or avoided entirely with change 2).
2. Avoids copying string names and values where possible. Previously string bytes would be copied from `data` to `_stringOctets` array, then copied from `_stringOctets` to `_headerNameOctets`/`_headerValueOctets`. Now a range is recorded of the start and end of a string value if it is completely available in `data` and it does not use huffman. Slices of original `data` are returned if the entire header was available.

Before:
```
|                                                 Method |        Mean |     Error |    StdDev |        Op/s | Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------------------------------------------- |------------:|----------:|----------:|------------:|------:|------:|------:|----------:|
|          DecodesLiteralHeaderFieldNeverIndexed_NewName |    111.2 ns |   2.26 ns |   2.22 ns | 8,993,376.5 |     - |     - |     - |         - |
|    DecodesLiteralHeaderFieldNeverIndexed_NewName_Large | 49,083.0 ns | 783.59 ns | 694.64 ns |    20,373.6 |     - |     - |     - |         - |
| DecodesLiteralHeaderFieldNeverIndexed_NewName_Multiple |    515.8 ns |  10.06 ns |  11.59 ns | 1,938,605.0 |     - |     - |     - |         - |
```

Avoiding state machine:
```
|                                                 Method |       Mean |    Error |   StdDev |        Op/s | Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------------------------------------------- |-----------:|---------:|---------:|------------:|------:|------:|------:|----------:|
|          DecodesLiteralHeaderFieldNeverIndexed_NewName |   103.6 ns |  2.08 ns |  2.23 ns | 9,656,957.8 |     - |     - |     - |         - |
|    DecodesLiteralHeaderFieldNeverIndexed_NewName_Large | 1,126.0 ns | 22.28 ns | 27.36 ns |   888,106.6 |     - |     - |     - |         - |
| DecodesLiteralHeaderFieldNeverIndexed_NewName_Multiple |   487.1 ns |  3.79 ns |  2.96 ns | 2,052,966.3 |     - |     - |     - |         - |
```

Avoiding state machine + string copies:
```
|                                                 Method |      Mean |    Error |   StdDev |         Op/s | Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------------------------------------------- |----------:|---------:|---------:|-------------:|------:|------:|------:|----------:|
|          DecodesLiteralHeaderFieldNeverIndexed_NewName |  71.90 ns | 1.163 ns | 1.088 ns | 13,908,049.2 |     - |     - |     - |         - |
|    DecodesLiteralHeaderFieldNeverIndexed_NewName_Large |  98.68 ns | 1.070 ns | 0.894 ns | 10,134,003.6 |     - |     - |     - |         - |
| DecodesLiteralHeaderFieldNeverIndexed_NewName_Multiple | 323.50 ns | 3.342 ns | 2.963 ns |  3,091,200.6 |     - |     - |     - |         - |
```

### CODE REVIEW

**Comment by stephentoub** [comment]
cc: @geoffkizer

**Comment by geoffkizer** [comment]
@JamesNK Does this move the needle for GRPC scenarios? If so, there are other potential optimizations we could consider: https://github.com/dotnet/runtime/issues/1505, https://github.com/dotnet/runtime/issues/1506

**Comment by JamesNK** [comment]
I haven't perf tested beyond micro-benchmarks.

I don't think this will have much impact with gRPC. gRPC doesn't use large headers, and most of the saving here is the time taken to parse string values.

Also, Kestrel supports writing dynamic HPack. SocketsHttpHandler probably receives headers as tiny indexes that are looked up from the dynamic table. The main winner from this change will be Kestrel parsing incoming headers from HttpClient (at least until HttpClient supports writing dynamic HPack).

**Comment by JamesNK** [comment]
Last thing

Previous changes:
```
|                                                 Method |      Mean |    Error |   StdDev |         Op/s | Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------------------------------------------- |----------:|---------:|---------:|-------------:|------:|------:|------:|----------:|
|          DecodesLiteralHeaderFieldNeverIndexed_NewName |  82.66 ns | 0.733 ns | 0.685 ns | 12,098,340.5 |     - |     - |     - |         - |
|    DecodesLiteralHeaderFieldNeverIndexed_NewName_Large | 109.39 ns | 1.440 ns | 1.347 ns |  9,141,350.2 |     - |     - |     - |         - |
| DecodesLiteralHeaderFieldNeverIndexed_NewName_Multiple | 335.33 ns | 2.813 ns | 2.493 ns |  2,982,159.7 |     - |     - |     - |         - |
|                 DecodesIndexedHeaderField_DynamicTable |  41.59 ns | 0.183 ns | 0.163 ns | 24,046,136.5 |     - |     - |     - |         - |
|        DecodesIndexedHeaderField_DynamicTable_Multiple | 162.23 ns | 0.977 ns | 0.914 ns |  6,164,196.6 |     - |     - |     - |         - |
```

Plus ref return HeaderValue struct from dynamic dictionary:
```
|                                                 Method |      Mean |    Error |   StdDev |         Op/s | Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------------------------------------------- |----------:|---------:|---------:|-------------:|------:|------:|------:|----------:|
|          DecodesLiteralHeaderFieldNeverIndexed_NewName |  83.39 ns | 0.694 ns | 0.649 ns | 11,992,273.6 |     - |     - |     - |         - |
|    DecodesLiteralHeaderFieldNeverIndexed_NewName_Large | 110.19 ns | 2.207 ns | 3.807 ns |  9,074,979.8 |     - |     - |     - |         - |
| DecodesLiteralHeaderFieldNeverIndexed_NewName_Multiple | 342.23 ns | 5.338 ns | 4.458 ns |  2,922,008.5 |     - |     - |     - |         - |
|                 DecodesIndexedHeaderField_DynamicTable |  36.82 ns | 0.682 ns | 0.638 ns | 27,160,947.4 |     - |     - |     - |         - |
|        DecodesIndexedHeaderField_DynamicTable_Multiple | 114.29 ns | 0.817 ns | 0.682 ns |  8,749,359.6 |     - |     - |     - |         - |
```

**Comment by JamesNK** [comment]
Need an approval before I can merge 🙏

**Review by halter73 — APPROVED** [review]

**Comment by halter73** [comment]
Sorry I didn't approve before preview7 branching. We can retarget the PR if need be.

File: src/Shared/runtime/Http2/Hpack/HPackDecoder.cs
Comment (JamesNK) [reply src/Shared/runtime/Http2/Hpack/HPackDecoder.cs:317]: That's the order the headers are defined in the spec.

File: src/Shared/runtime/Http2/Hpack/HPackDecoder.cs
Comment (JamesNK) [reply src/Shared/runtime/Http2/Hpack/HPackDecoder.cs:437]: Was already checked to ensure its size. I would update, but it is already mirrored 🤷

**Comment by JamesNK** [comment]
Fixes https://github.com/dotnet/aspnetcore/pull/23083

**Pull request merged**
